game.py

```diff
@@ -0,0 +1,302 @@
+from math import inf
+from copy import copy, deepcopy
+import random
+
+class GameState():
+    def __init__(self, board, player_next, previous_game_state = None):
+        self._board = board
+        self._player_next = player_next
+        self._previous_game_state = previous_game_state
+
+    def _copy(self):
+        board = deepcopy(self._board)
+        player_next = self._player_next
+        previous_game_state = deepcopy(self)
+        return GameState(board, player_next, previous_game_state)
+
+                
+    def get_board(self):
+        return self._board
+    
+    def get_player_next(self):
+        return self._player_next
+
+    def make_move(self, move):
+        new_game_state = self._copy()
+        new_game_state._board[move[0]][move[1]] = new_game_state._player_next
+        new_game_state._player_next = - new_game_state._player_next
+        return new_game_state
+        
+    def unmake_move(self):
+        new_game_state = self._previous_game_state
+        return new_game_state
+        
+    def get_next_moves(self, moves_sort = False, game_end = False):
+        moves = []
+        any_moves = False
+        for i in range(len(self._board)):
+            if any_moves:
+                break
+            for j in range(len(self._board[i])):
+                if self._board[i][j] == None:
+        #            moves.append ((i, j))
+                    moves.insert(0, (i, j))
+                    if game_end:
+                        any_moves = True
+                        break
+                    
+        random.shuffle(moves)
+        
+        if moves_sort:
+            moves_and_scores = []
+            i = 0
+            for move in moves:
+                new_game_state = self.make_move(move)
+                moves_and_scores.append([move])
+                moves_and_scores[i].append(new_game_state.score())
+                i = i + 1
+            print ('Moves and scores before sort : ', moves_and_scores)
+            moves_and_scores.sort(key = lambda x: x[1])
+            print ('Moves and scores after sort : ', moves_and_scores)
+            moves = []
+            for move_and_score in moves_and_scores:
+                moves.append(move_and_score[0])
+                
+        # print('moves are ', moves)
+        return moves
+
+    
+
+
+    def game_end(self):
+        if self.get_next_moves(game_end = True) == []:

```

Comment:
This reads a bit strangely to me and I personally would probably not try to optimize here (you're skipping the sorting-by-score already). If you were very keen to optimize, rather than complicating the `get_next_moves()` method I'd probably create another method like `is_any_move()`. There may be some duplicate-looking logic in that function, but overall less confusing to read.